### ISSUE TITLE
setpermissions.sh should not change the permissions for all files and folders in ~openhab

### ISSUE THREAD

**Issue by rkoshak** [openhab/openhab#3776]
The forum discussion is here: https://community.openhab.org/t/a-restart-of-openhab-causes-file-permissions-to-change-in-openhab-ssh/6121/7

Currently setpermissions.sh changes the permissions of all files and subfolders of ~openhab during an openhab restart. There are two problems with this.

1) It appears the intent is to set the permissions of the folders in /var/lib/openhab which include the Designer workspace, persistence, and zwave folders. However, if the user creates the openhab user themselves and maps the home directory to somewhere else, these folders get missed by setpermissions.sh

2) If the user wants to run ssh as the openhab (Exec binding) using ssh keys as opposed to a password (a very good security practice) they need to put those keys in ~openhab/.ssh. And the permissions on those ssh key files must be user rw only or else some (all?) ssh servers will reject the connection because the permissions on the private key are too open. Because the setpermissions.sh script blindly changes the permissions of all files and folders to 755 a restart of openhab will break ssh for the openhab user.

**Comment by theoweiss** [comment]
setpermissions is needed because the openhab execution user can be configured with /etc/default/openhab and it must be granted that this user has rw access to certain files.

I would suggest the following changes:
1) the OPENHAB_HOME_DIR value should be configurable in /etc/default/openhab
2) add a new variable to /etc/default/openhab SET_PERMISSIONS=yes|no with the default value "yes". If SET_PERMISSIONS is set to "no" no permission changes will happen.

**Comment by rkoshak** [comment]
If I'm understanding the changes correctly the changes are an all or nothing sort of change, correct? 1) lets you deal with issue 1) above where ~openhab isn't at /var/lib/openhab. 2) addresses the problem of it changing the .ssh directory permissions by essentially turning off the script entirely.

So my questions is if running setpermissions.sh every time OH starts is important, won't turning it off like this cause problems? If it isn't important, why is it run every time?

How hard would be it be to update the installer to put ~openhab somewhere else and change the script to operate on /var/lib/openhab instead of ${OPENHAB_HOME_DIR}.

Or make the script a little smarter to only change the permissions on those files or folders that are needed instead of blanket recursively update whole folders (at least for the ~openhab folder).

Or add a new variable to /etc/default/openhab to add exclusion patterns so certain folders like .ssh can be excluded.

**Comment by watou** [comment]
Would it not make sense to create the openhab user with a home directory in /home/openhab, so it becomes distinct from /var/lib/openhab?  In that case the 
```bash
for ohdir in "${OPENHAB_HOME_DIR}" "${OPENHAB_LOG_DIR}"; do
	setPermRecursive "${ohdir}"
done
```
never touch the actual openhab user's home directory, which must be left undisturbed by setpermissions.sh so it can host .ssh and perhaps other files and executable scripts (there was another report of a user's script losing its executable flag).

**Comment by watou** [comment]
@rkoshak, is this a new issue that appeared in 1.8.0?

**Comment by theoweiss** [comment]
I will work on this issue as soon as I'm ready with the openHAB2 deb packaging.

**Comment by rkoshak** [comment]
@watou, I think that this was also a problem with 1.7.1 but I didn't manage to find time to really look into it until after 1.8 was installed. I set up openHAB to execute a script that runs ssh back in 1.7.1 and I thought I got it to work. Then I went away for awhile and discovered a week or so later that it wasn't working after all but didn't get time to debug why until 1.8 was installed. User @Steamrunner on the forum reported that it was a problem for him as well on 1.7.1.
